File: clj/test/monkey_lang/lexer_test.clj
(ns monkey-lang.lexer-test
  (:require [clojure.test :refer [deftest is testing]]
            [monkey-lang.token :as token]
            [monkey-lang.lexer :as lexer]))

(def program 
  "
  let five = 5;
  let ten = 10;

  let add = fn(x, y) {
    x + y;
  };

  let result = add(five, ten);
  !-/*5;
  5 < 10 > 5;

  if (5 < 10) {
    return true;
  } else {
    return false;
  }

  10 == 10;
  10 != 9;
  \"foobar\"
  \"foo bar\"
  [1, 2];
  {\"foo\": \"bar\"}
  null
  let x = null
  ")

(deftest lexer-test
  (testing "=+(){},;"
    (let [result [[token/LET "let"]
                  [token/IDENT "five"]
                  [token/ASSIGN "="]
                  [token/INT "5"]
                  [token/SEMICOLON ";"]
                  [token/LET "let"]
                  [token/IDENT "ten"]
                  [token/ASSIGN "="]
                  [token/INT "10"]
                  [token/SEMICOLON ";"]
                  [token/LET "let"]
                  [token/IDENT "add"]
                  [token/ASSIGN "="]
                  [token/FUNCTION "fn"]
                  [token/LPAREN "("]
                  [token/IDENT "x"]
                  [token/COMMA ","]
                  [token/IDENT "y"]
                  [token/RPAREN ")"]
                  [token/LBRACE "{"]
                  [token/IDENT "x"]
                  [token/PLUS "+"]
                  [token/IDENT "y"]
                  [token/SEMICOLON ";"]
                  [token/RBRACE "}"]
                  [token/SEMICOLON ";"]
                  [token/LET "let"]
                  [token/IDENT "result"]
                  [token/ASSIGN "="]
                  [token/IDENT "add"]
                  [token/LPAREN "("]
                  [token/IDENT "five"]
                  [token/COMMA ","]
                  [token/IDENT "ten"]
                  [token/RPAREN ")"]
                  [token/SEMICOLON ";"]
                  [token/BANG "!"]
                  [token/MINUS "-"]
                  [token/SLASH "/"]
                  [token/ASTERISK "*"]
                  [token/INT "5"]
                  [token/SEMICOLON ";"]
                  [token/INT "5"]
                  [token/LT "<"]
                  [token/INT "10"]
                  [token/GT ">"]
                  [token/INT "5"]
                  [token/SEMICOLON ";"]
                  [token/IF "if"]
                  [token/LPAREN "("]
                  [token/INT "5"]
                  [token/LT "<"]
                  [token/INT "10"]
                  [token/RPAREN ")"]
                  [token/LBRACE "{"]
                  [token/RETURN "return"]
                  [token/TRUE "true"]
                  [token/SEMICOLON ";"]
                  [token/RBRACE "}"]
                  [token/ELSE "else"]
                  [token/LBRACE "{"]
                  [token/RETURN "return"]
                  [token/FALSE "false"]
                  [token/SEMICOLON ";"]
                  [token/RBRACE "}"]
                  [token/INT "10"]
                  [token/EQ "=="]
                  [token/INT "10"]
                  [token/SEMICOLON ";"]
                  [token/INT "10"]
                  [token/NOT_EQ "!="]
                  [token/INT "9"]
                  [token/SEMICOLON ";"]
                  [token/STRING "foobar"]
                  [token/STRING "foo bar"]
                  [token/LBRACKET "["]
                  [token/INT "1"]
                  [token/COMMA ","]
                  [token/INT "2"]
                  [token/RBRACKET "]"]
                  [token/SEMICOLON ";"]
                  [token/LBRACE "{"]
                  [token/STRING "foo"]
                  [token/COLON ":"]
                  [token/STRING "bar"]
                  [token/RBRACE "}"]
                  [token/NULL "null"]
                  [token/LET "let"]
                  [token/IDENT "x"]
                  [token/ASSIGN "="]
                  [token/NULL "null"]
                  [token/EOF ""]]]
    (is (= result (lexer/run program))))))

File: perl/lexer_test.pl
# use strict; # -- not enabled because the test will not work otherwise
use warnings;
use Test::More;
use lib 'Lexer';
use Lexer;
use Token;

sub test_nextToken {
    my ($input, $expectedTokens) = @_;

    my $l = Lexer->new($input);
    foreach my $expectedToken (@$expectedTokens) {
        my ($expectedType, $expectedLiteral) = @$expectedToken;
        my $tok = $l->nextToken();
        my $kind = $tok->getType();

        is($kind->getName(), $expectedType->getName(), "Token type matches");
        is($tok->getLiteral(), $expectedLiteral, "Token literal matches");
    }
}

my $code = '
    let five = 5;
    let ten = 10;
    let add = fn(x, y) {
        x + y;
    };
   let result = add(five, ten);
   !-/*5;
   5 < 10 > 5;
   if (5 < 10) {
       return true;
    } else {
        return false;
    }

    10 == 10;
    10 != 9;
';

test_nextToken($code,[
    [Token->LET, "let"],
    [Token->IDENT, "five"],
    [Token->ASSIGN, "="],
    [Token->INT, "5"],
    [Token->SEMICOLON, ";"],
    [Token->LET, "let"],
    [Token->IDENT, "ten"],
    [Token->ASSIGN, "="],
    [Token->INT, "10"],
    [Token->SEMICOLON, ";"],
    [Token->LET, "let"],
    [Token->IDENT, "add"],
    [Token->ASSIGN, "="],
    [Token->FUNCTION, "fn"],
    [Token->LPAREN, "("],
    [Token->IDENT, "x"],
    [Token->COMMA, ","],
    [Token->IDENT, "y"],
    [Token->RPAREN, ")"],
    [Token->LBRACE, "{"],
    [Token->IDENT, "x"],
    [Token->PLUS, "+"],
    [Token->IDENT, "y"],
    [Token->SEMICOLON, ";"],
    [Token->RBRACE, "}"],
    [Token->SEMICOLON, ";"],
    [Token->LET, "let"],
    [Token->IDENT, "result"],
    [Token->ASSIGN, "="],
    [Token->IDENT, "add"],
    [Token->LPAREN, "("],
    [Token->IDENT, "five"],
    [Token->COMMA, ","],
    [Token->IDENT, "ten"],
    [Token->RPAREN, ")"],
    [Token->SEMICOLON, ";"],
    [Token->BANG, "!"],
    [Token->MINUS, "-"],
    [Token->SLASH, "/"],
    [Token->ASTERISK, "*"],
    [Token->INT, "5"],
    [Token->SEMICOLON, ";"],
    [Token->INT, "5"],
    [Token->LT, "<"],
    [Token->INT, "10"],
    [Token->GT, ">"],
    [Token->INT, "5"],
    [Token->SEMICOLON, ";"],
    [Token->IF, "if"],
    [Token->LPAREN, "("],
    [Token->INT, "5"],
    [Token->LT, "<"],
    [Token->INT, "10"],
    [Token->RPAREN, ")"],
    [Token->LBRACE, "{"],
    [Token->RETURN, "return"],
    [Token->TRUE, "true"],
    [Token->SEMICOLON, ";"],
    [Token->RBRACE, "}"],
    [Token->ELSE, "else"],
    [Token->LBRACE, "{"],
    [Token->RETURN, "return"],
    [Token->FALSE, "false"],
    [Token->SEMICOLON, ";"],
    [Token->RBRACE, "}"],
    [Token->INT, "10"],
    [Token->EQ, "=="],
    [Token->INT, "10"],
    [Token->SEMICOLON, ";"],
    [Token->INT, "10"],
    [Token->NOT_EQ, "!="],
    [Token->INT, "9"],
    [Token->SEMICOLON, ";"],
    [Token->EOF, ""]
]);

done_testing();
